pkg/controller/machine/drain_controller.go

```diff
@@ -0,0 +1,163 @@
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	"k8s.io/kubectl/pkg/drain"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	machinev1 "github.com/openshift/api/machine/v1beta1"
+
+	"github.com/openshift/machine-api-operator/pkg/util/conditions"
+)
+
+// DrainController performs pods eviction for deleting node
+type machineDrainController struct {
+	client.Client
+	config *rest.Config
+	scheme *runtime.Scheme
+
+	eventRecorder record.EventRecorder
+}
+
+// newDrainController returns a new reconcile.Reconciler for machine-drain-controller
+func newDrainController(mgr manager.Manager) reconcile.Reconciler {
+	d := &machineDrainController{
+		Client:        mgr.GetClient(),
+		eventRecorder: mgr.GetEventRecorderFor("machine-drain-controller"),
+		config:        mgr.GetConfig(),
+		scheme:        mgr.GetScheme(),
+	}
+	return d
+}
+
+func (d *machineDrainController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	m := &machinev1.Machine{}
+	if err := d.Client.Get(ctx, request.NamespacedName, m); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.
+			return reconcile.Result{}, nil
+		}
+
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	existingDrainedCondition := conditions.Get(m, machinev1.MachineDrained)
+	alreadyDrained := existingDrainedCondition != nil && existingDrainedCondition.Status == corev1.ConditionTrue
+
+	if !m.ObjectMeta.DeletionTimestamp.IsZero() && *m.Status.Phase == phaseDeleting && !alreadyDrained {
+		drainFinishedCondition := conditions.TrueCondition(machinev1.MachineDrained)
+
+		if _, exists := m.ObjectMeta.Annotations[ExcludeNodeDrainingAnnotation]; !exists && m.Status.NodeRef != nil {
+			// pre-drain.delete lifecycle hook
+			// Return early without error, will requeue if/when the hook owner removes the annotation.
+			if len(m.Spec.LifecycleHooks.PreDrain) > 0 {
+				klog.Infof("%v: not draining machine: lifecycle blocked by pre-drain hook", m.Name)
+				d.eventRecorder.Eventf(m, corev1.EventTypeNormal, "DrainBlocked", "Drain blocked by pre-drain hook")
+				return reconcile.Result{}, nil
+			}
+			d.eventRecorder.Eventf(m, corev1.EventTypeNormal, "DrainProceeds", "Node drain proceeds")
+			if err := d.drainNode(ctx, m); err != nil {
+				klog.Errorf("%v: failed to drain node for machine: %v", m.Name, err)
+				conditions.Set(m, conditions.FalseCondition(
+					machinev1.MachineDrained,
+					machinev1.MachineDrainError,
+					machinev1.ConditionSeverityWarning,
+					"could not drain machine: %v", err,
+				))
+				d.eventRecorder.Eventf(m, corev1.EventTypeNormal, "DrainRequeued", "Node drain requeued: %v", err.Error())
+				return delayIfRequeueAfterError(err)
+			}
+			d.eventRecorder.Eventf(m, corev1.EventTypeNormal, "DrainSucceeded", "Node drain succeeded")
```

Comment:
There is an event at the end of `drainNode` which means we get two events every time this completes successfully, do we want to remove one of them?